hashing/sha256.cpp

```diff
@@ -0,0 +1,258 @@
+            uint32_t s0 =
+                right_rotate(a, 2) ^ right_rotate(a, 13) ^ right_rotate(a, 22);
+            uint32_t maj = (a & b) ^ (a & c) ^ (b & c);
+            uint32_t temp2 = s0 + maj;
+
+            h = g;
+            g = f;
+            f = e;
+            e = d + temp1;
+            d = c;
+            c = b;
+            b = a;
+            a = temp1 + temp2;
+        }
+
+        // Update hash values
+        hash[0] += a;
+        hash[1] += b;
+        hash[2] += c;
+        hash[3] += d;
+        hash[4] += e;
+        hash[5] += f;
+        hash[6] += g;
+        hash[7] += h;
+    }
+
+    return hash;
+}
+
+/**
+ * @brief Convert the hash to a hexadecimal string
+ * @param hash Hash array
+ * @return std::string Final hash value
+ */
+std::string hash_to_string(const std::array<uint32_t, 8> &hash) {
+    std::string result;
+    for (size_t i = 0; i < 8; ++i) {
+        for (size_t j = 0; j < 4; ++j) {
+            uint32_t byte = (hash[i] >> (24 - j * 8)) & 0xFF;
+            std::stringstream ss;
+            ss << std::hex << std::setfill('0') << std::setw(2) << byte;
+            result += ss.str();
+        }
+    }
+
+    return result;
+}
+
+/**
+ * @brief The SHA-256 algorithm
+ * @param input The input string to hash
+ * @return std::string The final hash value
+ */
+std::string sha256(const std::string &input) {
+    std::array<uint32_t, 8> hash = compute_hash(input);
+
+    std::string result = hash_to_string(hash);
+
+    return result;
+}
+}  // namespace sha256
+}  // namespace hashing
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    struct TestCase {
+        const std::string input;
+        const std::string expected_hash;
+        TestCase(std::string input, std::string expected_hash)
+            : input(std::move(input)),
+              expected_hash(std::move(expected_hash)) {}
+    };
+    const std::vector<TestCase> test_cases{
+        TestCase(
+            "",
+            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"),
+        TestCase(
+            "test",
+            "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"),
+        TestCase(
+            "Hello World",
+            "a591a6d40bf420404a011733cfb7b190d62c65bf0bcda32b57b277d9ad9f146e"),
+        TestCase("Hello World!",
+                 "7f83b1657ff1fc53b92dc18148a1d65dfc2d4b1fa3d677284addd200126d9"
+                 "069")};
+    for (const auto &tc : test_cases) {
+        assert(hashing::sha256::sha256(tc.input) == tc.expected_hash);
+    }
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // Run self-test implementations
+
+    // Custom Input
+    std::string input;
+    std::cout << "Enter a message to be hashed: ";
+    getline(std::cin, input);
+    std::string hash = hashing::sha256::sha256(input);
+    std::cout << "SHA-256 Hash: " << hash << std::endl;
```

Comment:
I am not sure if this is [really needed](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#test). Maybe move the entire `hashing` namespace to a header (and corresponding `cpp` file), and provide two files likes `test_sha256.cpp` and `example_sha256.cpp`.

Reply:
Ok then I'll remove this part and just print a confirmation message when tests have passed successfully.

Reply:
> Ok then I'll remove this part and just print a confirmation message when tests have passed successfully.

Strictly speaking, even this is not needed: if some of the tests fail, the program has a non-zero exit code.